File: python_code/graph/direct/TopologySortBFS.py
from collections import deque

from graph.direct.GraphImpl import GraphImpl


class TopologySortBFS:
    def __init__(self, g: GraphImpl):
        if not g.isDirectedGraph():
            raise Exception("只能对有向图进行拓扑排序")

        self.hasCycle = False
        indegrees = [0] * g.getV()
        for v in range(g.getV()):
            indegrees[v] = g.indegree(v)

        queue = deque()
        for v in range(g.getV()):
            if indegrees[v] == 0:
                queue.append(v)

        self.res = [0] * g.getV()
        index = 0
        while len(queue):
            v = queue.popleft()
            self.res[index] = v
            index += 1
            for w in g.adj(v):
                indegrees[w] -= 1
                if indegrees[w] == 0:
                    queue.append(w)

        if index != g.getV():
            self.hasCycle = True

    def isHasCycle(self):
        return self.hasCycle

    def getRes(self):
        return self.res

if __name__ == '__main__':
    g = GraphImpl("C:/workspace/douma-algo/data/directedgraph-dfs.txt", True)
    bfs = TopologySortBFS(g)
    print(bfs.isHasCycle())
    print(bfs.getRes())

File: javascript_code/graph/direct/TopologySortBFS.js


export class TopologySortBFS {

    constructor(g) {
        if (!g.getIsDirected()) {
            throw new Error("只能对有向图进行拓扑排序")
        }
        this.g = g
        this.hasCycle = false

        const indegrees = new Array(g.getV()).fill(0)
        for (let v = 0; v < g.getV(); v++) {
            indegrees[v] = g.indegree(v)
        }

        const queue = []
        for (let v = 0; v < g.getV(); v++) {
            if (indegrees[v] == 0) {
                queue.push(v)
            }
        }

        this.res = new Array(g.getV()).fill(0)
        let index = 0
        while (queue.length) {
            const v = queue.shift()
            this.res[index++] = v
            for (const w of g.adj(v)) {
                indegrees[w]--
                if (indegrees[w] == 0) {
                    queue.push(w)
                }
            }
        }
        if (index != g.getV()) {
            this.hasCycle = true
        }
    }

    isHasCycle() {
        return this.hasCycle
    }

    getRes() {
        return this.res
    }
}